lib/Transforms/Scalar/DxilFixConstArrayInitializer.cpp

```diff
@@ -0,0 +1,184 @@
+#include "llvm/Pass.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/GlobalVariable.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Operator.h"
+#include "llvm/IR/CFG.h"
+#include "llvm/Transforms/Scalar.h"
+#include "dxc/DXIL/DxilModule.h"
+#include "dxc/HLSL/HLModule.h"
+
+#include <unordered_map>
+#include <limits>
+
+using namespace llvm;
+
+namespace {
+
+class DxilFixConstArrayInitializer : public ModulePass {
+public:
+  static char ID;
+  DxilFixConstArrayInitializer() : ModulePass(ID) {
+    initializeDxilFixConstArrayInitializerPass(*PassRegistry::getPassRegistry());
+  }
+  bool runOnModule(Module &M) override;
+  const char *getPassName() const override { return "Dxil Fix Const Array Initializer"; }
+};
+
+char DxilFixConstArrayInitializer::ID;
+}
+
+static bool TryFixGlobalVariable(GlobalVariable &GV, BasicBlock *EntryBlock, const std::unordered_map<Instruction *, unsigned> &InstOrder) {
+  // Only proceed if the variable has an undef initializer
+  if (!GV.hasInitializer() || !isa<UndefValue>(GV.getInitializer()))
+    return false;
+
+  // Only handle cases when it's a scalar type or if it's an array of
+  // scalars.
+  Type *Ty = GV.getType()->getPointerElementType();
+  if (Ty->isArrayTy() && Ty->getArrayElementType()->isArrayTy())
+    return false;
+
+  // Record the earliest load and latest store in the entry block.
+  unsigned EarliestLoad = std::numeric_limits<unsigned>::max();
+  unsigned LatestStore = 0;
+
+  bool IsArray = Ty->isArrayTy();
+  SmallVector<Constant *, 16> InitValue;
+  SmallVector<unsigned, 16> LatestStores;
+
+  Type *ElementTy = nullptr;
+  if (IsArray) {
+    InitValue.resize(Ty->getArrayNumElements());
+    LatestStores.resize(Ty->getArrayNumElements());
+    ElementTy = Ty->getArrayElementType();
+  }
+  else {
+    InitValue.resize(1);
+    LatestStores.resize(1);
+    ElementTy = Ty;
+  }
+
+  SmallVector<StoreInst *, 8> StoresToRemove;
+
+  for (User *U : GV.users()) {
+    if (GEPOperator *GEP = dyn_cast<GEPOperator>(U)) {
+      bool AllConstIndices = GEP->hasAllConstantIndices();
+      unsigned NumIndices = GEP->getNumIndices();
+      if (IsArray && NumIndices != 2)
+        return false;
+      if (!IsArray && NumIndices != 1)
+        return false;
+
+      for (User *GEPUser : GEP->users()) {
+        if (StoreInst *Store = dyn_cast<StoreInst>(GEPUser)) {
+          if (Store->getParent() == EntryBlock) {
+            if (!isa<Constant>(Store->getValueOperand()))
```

Comment:
(consider doing a `dyn_cast` once rather an `isa` and a later disconnected `cast`)